cupy/core/core.pyx

```diff
@@ -1565,6 +1565,41 @@ cdef class ndarray:
     # TODO(okuta): Implement __setslice__
     # TODO(okuta): Implement __contains__
 
+    # numpy/ufunc compat
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+
+        """Apply unary or binary ufunc to this array
+
+        If binary, only allow if second argument is another cupy ndarray or
+        a number, i.e., raise ValueError instead of silently converting a
+        numpy array.
+        """
+        import cupy as cp  # top-level ufuncs
+        if 'out' in kwargs:
+            # need to unfold tuple argument in kwargs
+            check = inputs + kwargs['out']
+            kwargs['out'] = kwargs['out'][0]
+        else:
+            check = inputs
+
+        name = ufunc.__name__
+        if method == '__call__':
+            if ufunc.signature is not None:
+                # we don't support generalised-ufuncs (gufuncs)
+                return NotImplemented
+            try:
+                cp_ufunc = getattr(cp, ufunc.__name__)
+                return cp_ufunc(*inputs, **kwargs)
+            except AttributeError:
+                return NotImplemented
+        # Don't use for now, interface uncertain
+        # elif method =='at' and name == 'add':
+            # the only ufunc attribute currently
+            # http://docs-cupy.chainer.org/en/stable/reference/ufunc.html#ufunc-at
+            # self.scatter_add(*inputs, **kwargs)
+        else:
```

Comment:
Do we have a chance to support some of the reduction routines (like `add.reduce` via `sum`, `maximum.reduce` via `amax`, etc.) here (maybe via ad-hoc manual mapping)? Or do you think it's better to first support `reduce` in CuPy's ufuncs (so that no ad-hoc coding is needed here)?

Reply:
Note that numpy reductions already dispatch if an appropriate method is defined

```python
In [1]: class Foo:
   ...:     def sum(*args, **kwargs):
   ...:         return 123
   ...:     

In [2]: import numpy as np

In [3]: np.sum(Foo())
Out[3]: 123
```

Adding support for ufunc.reduce would also be good, but this seems better to do in the ufuncs themselves rather than in the `__array_ufunc__` protocol.

Reply:
Ah, sum is already dispatched. I misunderstood.